### ISSUE TITLE
Terser fails to minimize comparation with object

### ISSUE THREAD

**Issue by Tal500** [terser/terser#1335]
A bug/question: Why Terser fails to minimize comparation with object? I'll copy the example shot from Terser REPL:

Terser options:
```js
// edit terser options
{
  module: true,
  compress: {},
  mangle: {},
  output: {},
  parse: {},
  rename: {},
}
```
Input:
```js
// write or paste code here

console.log(`1: ${false === true}`);
console.log(`2: ${false === null}`);
console.log(`3: ${true === null}`);
console.log(`4: ${{a:7} === null}`);
console.log(`5: ${{a:7} === undefined}`);
console.log(`6: ${{a:7} === false}`);
console.log(`7: ${{a:7} === true}`);
console.log(`8: ${{a:7} == {a:7}}`);
console.log(`9: ${{a:7} === {a:6}}`);
const obj = {a:7};
console.log(`10: ${obj === obj}`);
console.log(`11: ${obj === true}`);
```

Output:
```js
console.log("1: false"),console.log("2: false"),console.log("3: false"),console.log(`4: ${null==={a:7}}`),console.log(`5: ${void 0==={a:7}}`),console.log(`6: ${!1==={a:7}}`),console.log(`7: ${!0==={a:7}}`),console.log(`8: ${{a:7}=={a:7}}`),console.log(`9: ${{a:7}==={a:6}}`);const o={a:7};console.log("10: true"),console.log(`11: ${!0===o}`);
```

Why do we have that except for the case of
```js
const obj = {a:7};
console.log(`10: ${obj === obj}`);
```
, every comparation with object can't be reduced, but why?

## Reason I need it
I'm trying to use the Vite `import.meta.env` metadata, but I want that if someone will build the code without Vite, no error will happen to him. Specifically:
```js
export const browser = import.meta.env != null && import.meta.env.SSR != null ? !import.meta.env.SSR : typeof window !== 'undefined'
```

But because what I said, even on Vite that defines this props, it is not get minimized, which doesn't allow me to enjoy dead code optimization on `if (browser) {}` for example.

## Extra note
This issue happen to me also with ESBuild minification, so I think there might be a good internal-JS reason for this behavior, but I couldn't find docs why.

**Comment by Tal500** [comment]
# Update for my use case

I've found out that changing the `browser` definition to this one does get optimized well to `true` or `false`:
```js
const browser = import.meta.env && import.meta.env.SSR != null ? !import.meta.env.SSR : typeof window !== 'undefined';
```
(the difference is that before I checked `import.meta.env != null`)

However:
1. The issue remains, though I've found a workaround.
2. _Side note_: ESBuild minification didn't manage to optimize this updated code. Good job you guys at Terser!

**Comment by fabiosantoscode** [comment]
How did you get Terser to optimize that to true or false? I'm assuming you used the define option perhaps?

**Comment by Tal500** [comment]
> How did you get Terser to optimize that to true or false? I'm assuming you used the define option perhaps?

At first I gave a clean reproduction for you to check. After, I told my real world use case.
In this use case, I'm using the Vite build system. Vite define and replaces the `import.meta.env` and `import.meta.env.SSR`, and when production build, only after bundling via Rollup (Vite uses rollup internally for production builds), Vite invokes `esbuild` or `terser` output code minification, depending on the configuration.

So basically, yes, you can think I used define for them.
